### DISCUSSION TITLE
[开发]如何VSCode中调试插件？

### DISCUSSION THREAD

**Discussion by Andy-AO** [bookfere/Ebook-Translator-Calibre-Plugin#266] [Q&A]
各位此前的贡献者们，你们是如何在IDE中调试插件的，求指教！

在插件中插入`from calibre.rpdb import set_trace; set_trace(port=5678)`，然后运行`calibre-debug -g`，可以发现pdb成功监听。
使用`calibre-debug -c "from calibre.rpdb import cli; cli(5678)"`可以在命令行中进入pdb

![image](https://github.com/bookfere/Ebook-Translator-Calibre-Plugin/assets/38101170/10cfb529-3322-4adf-8f2e-a78b1bc230e7)

可是如何在VSCode中使用pdb进行调试？

尝试配置launch.json，虽然运行成功，但之后UI上的debug操作都是灰色的，没法操作

```JavaScript
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
       {
            "name": "Python Debugger: Attach",
            "type": "debugpy",
            "request": "attach",
            "connect": {
              "host": "127.0.0.1",
              "port": 5678
            },
            "justMyCode": false
          }
    ]
}
```

![image](https://github.com/bookfere/Ebook-Translator-Calibre-Plugin/assets/38101170/cc68de6f-6f16-4c54-a0b0-6958ffc10558)

**Comment by Andy-AO** [comment]
这个配置也不行，UI上没有反应

```JavaScript
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [
      {
          "name": "TestDebug：Python_attach",
          "type": "python",
          "request": "attach",
          "port": 5678,
          "host": "localhost",
          "justMyCode": false
      }
  ]
}
```
![image](https://github.com/bookfere/Ebook-Translator-Calibre-Plugin/assets/38101170/50f100ba-3a6d-42fb-bce5-3c7d3a83f174)

**Comment by bookfere** [comment]
Calibre，包括其插件，都没很好的支持 IDE 调试器。在插件开发文档的 [Debugging plugins](https://manual.calibre-ebook.com/creating_plugins.html#debugging-plugins) 小节中有这样一段描述：

> You can insert print statements anywhere in your plugin code, they will be output in debug mode. Remember, this is Python, you really shouldn’t need anything more than print statements to debug ;) I developed all of calibre using just this debugging technique.

Calibre 作者在论坛的帖子中（[1](https://www.mobileread.com/forums/showthread.php?t=345343)，[2](https://www.mobileread.com/forums/showthread.php?t=325916)，[3](https://www.mobileread.com/forums/showthread.php?t=345343)）也多次提到他从不使用 IDE 调试器，而是用简单的 `print()` 调试。

不过针对调试 Calibre 的文档中 [Using the debugger in your favorite Python IDE](https://manual.calibre-ebook.com/develop.html#id16) 这一小节倒是简略提到了支持 IDE 调试器。

不过我在测试以后发现这种调试方式不太可行。有些帖子（[1](https://www.mobileread.com/forums/showthread.php?t=251201)、[2](https://stackoverflow.com/questions/64715146/debugging-python-in-vscodium-with-attach-using-process-id-times-out-on-macos-c)、[3](https://www.mobileread.com/forums/showthread.php?t=334662)）也有结合其他 IDE 使用调试器的讨论，但好像也都不太可行。

**Comment by bookfere** [comment]
这里有一个 Calibre 插件作者提供的[插件开发指南](https://github.com/kiwidude68/calibre_plugins/blob/main/CONTRIBUTING.md)也可以做一个参考。
